### PR TITLE
[controls] remove Linq usage and collection copying

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Reviewing `dotnet trace` output of `dotnet new maui`:

2.64ms System.Linq!System.Linq.Enumerable.Where
    ... ObservableWrapper<TTrack_ref,TRestrict_REF>.GetEnumerator

2.61ms System.Linq!System.Linq.Enumerable.FirstOrDefault
    ... TypeConversionExtensions.GetTypeConverterTypeName

2.81ms System.Linq!System.Linq.Enumerable.ToList
    ... IShellController.GetItems
    ... IVisualTreeElement.GetVisualChildren

I reworked each of these places to just use plain `foreach` loops, and
not copy entire collections where appropriate.

`ObservableWrapper` could just use the underlying collection for many
of the methods, where we shouldn't need our own implementation.

### CODE REVIEW

**Review by Redth — APPROVED** [review]

**Comment by jonathanpeppers** [comment]
I need to fix some tests here.

File: src/Controls/src/Core/Shell/ShellItem.cs [hunk 1]
```diff
@@ -328,6 +327,6 @@ internal override void SendDisappearing()
 			}
 		}
 
-		IReadOnlyList<Maui.IVisualTreeElement> IVisualTreeElement.GetVisualChildren() => Items.ToList().AsReadOnly();
+		IReadOnlyList<Maui.IVisualTreeElement> IVisualTreeElement.GetVisualChildren() => Items.ToList();
```
Comment (jonathanpeppers) [inline src/Controls/src/Core/Shell/ShellItem.cs:330]: This avoids an allocation. `.ToList()` already creates a unique copy of the list, so we can just return it as-is.

https://source.dot.net/#System.Private.CoreLib/List.cs,246

File: src/Controls/src/Core/ObservableWrapper.cs [hunk 2]
```diff
@@ -106,7 +107,13 @@ IEnumerator IEnumerable.GetEnumerator()
 
 		public IEnumerator<TRestrict> GetEnumerator()
 		{
-			return _list.Where(i => i.Owned).OfType<TRestrict>().GetEnumerator();
+			foreach (TRestrict item in _list)
```
Comment (eerhardt) [inline src/Controls/src/Core/ObservableWrapper.cs:110]: Does this cause a cast exception when you have something in `_list` that isn't a `TRestrict`?

File: src/Controls/src/Core/ObservableWrapper.cs [hunk 3]
```diff
@@ -291,10 +298,7 @@ public void Remove(object value)
 			Remove((TRestrict)value);
 		}
 
-		public void CopyTo(Array array, int index)
-		{
-			CopyTo(array.Cast<TRestrict>().ToArray(), index);
```
Comment (eerhardt) [inline src/Controls/src/Core/ObservableWrapper.cs:296]: This code didn't work before, right? Wasn't this copying to a new array which then just got discarded?

File: src/Controls/src/Core/ObservableWrapper.cs [hunk 4]
```diff
@@ -291,10 +298,7 @@ public void Remove(object value)
...
-		}
+		public void CopyTo(Array array, int index) => _list.CopyTo((TTrack[])array, index);
```
Comment (eerhardt) [inline src/Controls/src/Core/ObservableWrapper.cs:301]: I think you need to call the CopyTo method on line 57. That way only the `TRestrict` items get copied into the array.

File: src/Controls/src/Core/ObservableWrapper.cs (hunk 2 above)
Comment (jonathanpeppers) [reply src/Controls/src/Core/ObservableWrapper.cs:110]: I figured out a test that breaks, let me fix these.

File: src/Controls/src/Core/ObservableWrapper.cs (hunk 3 above)
Comment (jonathanpeppers) [reply src/Controls/src/Core/ObservableWrapper.cs:296]: Same here, I think I can write a test that breaks.

File: src/Controls/src/Core/Shell/ShellItem.cs [hunk 5]
```diff
@@ -84,8 +84,7 @@ bool IShellItemController.ProposeSection(ShellSection shellSection, bool setValu
 		}
 
 		// we want the list returned from here to remain point in time accurate
-		ReadOnlyCollection<ShellSection> IShellItemController.GetItems() =>
-			new ReadOnlyCollection<ShellSection>(((ShellSectionCollection)Items).VisibleItemsReadOnly.ToList());
+		ReadOnlyCollection<ShellSection> IShellItemController.GetItems() => ((ShellSectionCollection)Items).VisibleItemsReadOnly;
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/Controls/src/Core/Shell/ShellItem.cs:87]: @PureWeen do you know the details why this list needs to be copied?

[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Controls/src/Core/Shell/ShellItem.cs:87]: This relates to what we chatted about today. The point was to make these all point in time valid for the platform so things wouldn't change from under anyone's feet. Most of the test code for this is part of our Legacy Gallery which we can't really run right now unfortunately. The main area that flexes this concept would be toggling "IsVisible" on various shell elements (especially a currently visible one) and making sure the platform doesn't crash. I have a few tests for this in the Legacy Gallery.

I'd say we can merge this one now and once I get the legacy tests up and running then I'm going to do more of an exhaustive review for regressions.

[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/Controls/src/Core/Shell/ShellItem.cs:87]: I don't think what changed here will break -- but we'd like to try removing _more_ collection copying.

So when we get those tests back, we can try that.

**Pull request merged**